### PR TITLE
Wildcard subscription fix

### PR DESCRIPTION
**Pull request by sheindel**
Resolves https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/issues/84. 

Currently, the SDK does not correctly match wildcard topics. This should resolve this issue (as well as others). Test cases included.

### CODE REVIEW

File: src/libiotc/iotc_user_sub_call_wrapper.c [hunk 1]
```diff
@@ -73,7 +73,7 @@ iotc_state_t iotc_user_sub_call_wrapper(void* context, void* data,
           msg->publish.content ? msg->publish.content->data_ptr : NULL;
       params.message.temporary_payload_data_length =
           msg->publish.content ? msg->publish.content->length : 0;
-      params.message.topic = (const char*)sub_data->subscribe.topic;
+      params.message.topic = (const char*)msg->publish.topic_name->data_ptr;
```
Comment (sheindel) [inline src/libiotc/iotc_user_sub_call_wrapper.c]: This puts the published topic into the callback handler so callbacks looking for wildcards can get the subfolder topic

**Comment by atigyi** [comment]
@sheindel, I've added a batch unit test case, two cases are failing. Can we figure out an elegant solution to handle the subscription/publish topic pairs in question? Thanks!

File: src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c [hunk 2]
```diff
@@ -254,6 +254,61 @@ IOTC_TT_TESTCASE(utest__match_topics__valid_data__match_topics_should_return_1,
...
+IOTC_TT_TESTCASE(utest__match_topics__batch_tests, {
+  iotc_event_handle_t handle = iotc_make_empty_handle();
+
+  typedef struct {
+    char* subscription_topic;
+    const char* published_topic;
+    const uint8_t expected_result;
+  } iotc_match_topics_test_cases_t;
+
+  iotc_match_topics_test_cases_t test_cases[] = {
```
Comment (sheindel) [inline src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c]: @atigyi I really appreciate this format, are we okay to modify all topic match test cases to this format? I was just sticking with what was in place but I'd be open to refactoring

File: src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c [hunk 3]
```diff
@@ -77,44 +77,238 @@ static iotc_state_t failed_subscribe_handler(
...
+    {"long_topic_name_same_length_1", "long_topic_name_same_length_2", 1},
+    {"long_topic_name_same_length_2", "long_topic_name_same_length_1", 1},
+    {"long_topic_name", "long_topic_name_different_length", 1},
+    {"long_topic_name_different_length", "short", 1},
+    {NULL, "short", 1},
+    {"sub", NULL, 1},
+    {NULL, NULL, 1},
+
+    {"multi/level/#", "multi/level", 0},
+    {"multi/level/#", "multi/level/", 0},
```
Comment (sheindel) [inline src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c]: I see what is failing, but I think I also have a bug here and your test case is wrong. I think this actually should match and my test fails these, so this should be 1.

File: src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c [hunk 4]
```diff
@@ -77,44 +77,238 @@ static iotc_state_t failed_subscribe_handler(
...
+    {"multi/level/#", "multi/level/topic", 0},
+    {"multi/level/#", "multi/level/topic/", 0},
```
Comment (sheindel) [inline src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c]: @atigyi same here, my code breaks on this and it should match

File: src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c (hunk 2 above)
Comment (atigyi) [reply src/tests/utests/iotc_utest_mqtt_logic_layer_subscribe.c]: :+1:

**Comment by atigyi** [comment]
@sheindel, thanks for the updates! Looks good! I just pushed a change, please review it. Also ran formatting: `clang-format -i -style=Google iotc_mqtt_logic_layer_data_helpers.h`.
Is you are ok with this last change I am ok to merge it to development branch!

**Comment by sheindel** [comment]
> @sheindel, thanks for the updates! Looks good! I just pushed a change, please review it. Also ran formatting: `clang-format -i -style=Google iotc_mqtt_logic_layer_data_helpers.h`.
> If you are ok with this last change I am ok to merge it to development branch!

@atigyi Look great to me. Thanks for the `clang-format` info as well, I will use that if I have the privilege to contribute in the future.

**Pull request merged**